src/app/digital/public/api/impl/DigitalComponent.ts

```diff
@@ -16,4 +18,34 @@ export class DigitalComponentImpl extends ComponentImpl<
     public override get info(): DigitalComponentInfo {
         return new DigitalComponentInfoImpl(this.circuit, this.kind);
     }
+
+    public override firstAvailable(portGroup: Port["group"]): DigitalPort | undefined {
+        const ports = this.internal.getPortsForComponent(this.id);
+
+        const inputTypes = this.info.inputPortGroups;
+        const outputTypes = this.info.outputPortGroups;
+
+        const isInputType = inputTypes.includes(portGroup);
+        const isOutputType = outputTypes.includes(portGroup)
+
+        if (!isInputType && !isOutputType) 
+            return undefined;
+
+        function firstAvailableHelper(comp: DigitalComponentImpl, id: string): Boolean {
+            let portObject = comp.internal.getPortByID(id);
+            let portWire = comp.internal.getWiresForPort(id); // no wires = available
+    
+            if (comp.info.inputPortGroups && portObject !== undefined && portObject.group === portGroup)
```

Comment:
Checking `if (comp.info.inputPortGroups)` would just check if the array is defined or not but it should always be defined so I think it's a no-op currently

if you're trying to see if it's *in` the inputPortGroups, you gotta do `.includes(portObject.group)`